### ISSUE TITLE
Switch between light/dark UI colors and day/night map style depending on time of day

### ISSUE THREAD

**Issue by 1ec5** [mapbox/mapbox-navigation-android#4660]
The map view should automatically switch between the Mapbox Navigation Day and Mapbox Navigation Night map styles (or corresponding styles specified by the developer) based on the time of day and also when entering or exiting a tunnel. Ideally, foreground and background colors throughout the built-in UI components should also change to match the map style. For this reason, the ability to switch map styles should be the responsibility of UI code in the navigation SDK, built on top of map SDK functionality but not built into the map SDK.

The iOS navigation SDK calls this customization system “[styles](https://docs.mapbox.com/ios/beta/navigation/guides/app-styling/)”, though we sometimes refer to it as “UI styles” or “themes” or “skins” to differentiate it from the more basic style switching functionality exposed by the map SDK.

/ref #1057 #3423
/cc @mapbox/navigation-android

**Comment by 1ec5** [comment]
There are some examples of what the iOS navigation SDK looks like by day and by night in mapbox/mapbox-navigation-ios#2291. The most glaring difference is that the top and bottom banners’ backgrounds are white during the day and only midnight blue at night. That distinction allows the iOS navigation SDK to look like a plain-vanilla UI on iOS, but it may be contrary to design expectations on Android. The important thing is that there is the capability to vary the colors of more than just the map.

**Comment by cafesilencio** [comment]
Is there a developer override or does the SDK do this switching whether the developer wants it or not? How does this work with custom styles, for example if a developer creates only one custom map style, how does the SDK switch to the "dark" version?

**Comment by cafesilencio** [comment]
> It should not be up to the application to switch the map style directly.

I disagree. The SDK shouldn't impose such conditions onto a developer IMO. They have their own branding and app. requirements which we cannot possibly predict.

**Comment by abhishek1508** [comment]
I agree with the idea of changing the style between day/night based on the time. We could also look into LUX value coming out of the device sensors and hook the style changing to it. 
> Is there a developer override or does the SDK do this switching whether the developer wants it or not?

I definitely think this kind of override should be there

> for example if a developer creates only one custom map style, how does the SDK switch to the "dark" version

The SDK continues to show day style in dark mode.

> I disagree. The SDK shouldn't impose such conditions onto a developer IMO. 

I agree with @cafesilencio. We shouldn't impose such kind of restrictions. But both the behaviors should not co-exist as it would increasingly complicate the logic and doesn't make sense. If a developer wants to decide when to change the styles they should first disable auto switching.

**Comment by 1ec5** [comment]
> Is there a developer override or does the SDK do this switching whether the developer wants it or not? How does this work with custom styles, for example if a developer creates only one custom map style, how does the SDK switch to the "dark" version?

The default is to switch between Mapbox Navigation Day and Mapbox Navigation Night automatically, but the developer can always customize these two slots or define only a single style, which disables the day/night switching functionality.

>> It should not be up to the application to switch the map style directly.
> 
> I disagree. The SDK shouldn't impose such conditions onto a developer IMO. They have their own branding and app. requirements which we cannot possibly predict.

I should’ve worded this differently: it should always be possible for the application to change the map style, but something in the navigation SDK should manage the lifecycle of the style, since the map SDK doesn’t have context about the day/night switching. It also isn’t purely about day and night: a night style is automatically applied when entering a tunnel, something the map SDK couldn’t possibly know on its own.

On iOS, StyleManager is a reusable component that the developer can attach to any NavigationMapView. The one built into the drop-in UI has one already provided. If you provide only one UI style to StyleManager, it does nothing. I want to also emphasize that the UI styles themselves can be customized. We encourage any developer who wants a custom look and feel to define a custom UI style that specifies UI colors that match a custom map style.

**Comment by artemsorochan** [comment]
Hi, was this feature implemented?
